### ISSUE TITLE
RootDirectory fails maybe on some (or all) options that create a mount namespace

### ISSUE THREAD

**Issue by evverx** [systemd/systemd#4588]
Based on https://github.com/systemd/systemd/pull/4533#issuecomment-258327773

```
[Service]
ExecStart=/busybox id
RootDirectory=/chroot
PrivateTmp=yes
```

```
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Trying to enqueue job hola.service/start/replace
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Installed new job hola.service/start as 194
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Enqueued job hola.service/start as 194
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Passing 0 fds to service
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: About to execute: /busybox id
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Forked /busybox as 197
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Changed dead -> running
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Job hola.service/start finished, result=done
Nov 04 04:40:06 systemd-testsuite systemd[1]: Started hola.service.
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Child 197 belongs to hola.service
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Main process exited, code=exited, status=226/NAMESPACE
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Changed running -> failed
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Unit entered failed state.
Nov 04 04:40:06 systemd-testsuite systemd[1]: hola.service: Failed with result 'exit-code'.
```
Also I checked: `ProtectKernelTunables`, `DynamicUsers`

BTW: there is #4567 . But that's about the `chase_symlinks`-function really

Seems like this worked ee818b89f4890b3a00e9. So, I marked as "regression" too

**Comment by evverx** [comment]
I've checked `v229`:

```
Started hola.service.
hola.service: Main process exited, code=exited, status=203/EXEC
hola.service: Unit entered failed state.
hola.service: Failed with result 'exit-code'.
Started hola.service.
```

So, this doesn't work :-) Removed the "regression"-tag

Actually, I should

```
mkdir /chroot/tmp
mkdir -p /chroot/var/tmp
```

But that's not so obvious. I guess we should document this at least.

**Comment by tixxdz** [comment]
@evverx Indeed.

Now to sum it up: https://github.com/systemd/systemd/pull/4594 and https://github.com/systemd/systemd/pull/4596 fixes RootDirectory= and make it possible to use it with other options. Like ProtectHome= ProtectKernelModules= ProtectKernelTunables= even if the paths are missing and others like PrivateTmp etc only if you create the directories manually.

However the only limit is DynamicUser=yes and ProtectSystem=full|strict see PR https://github.com/systemd/systemd/pull/4596 why. Maybe it may fail to other reasons, but currently making  RootDirectory= and DynamicUser= work together seems not that trivial, and we need more feedback, so postponed.

cc @poettering

**Comment by poettering** [comment]
Fix waiting in #4727.

**Comment by evverx** [comment]
https://github.com/systemd/systemd/pull/4727#issuecomment-266901501
> force pushed a new version that drops the RootDirectory= patch

**Comment by poettering** [comment]
@evverx hmm, what is this bug precisely about? does it still apply to current git? /me has trouble grokking what was meant here...

**Comment by evverx** [comment]
@poettering , it's about documentation. It's not clear how to use different features together. For example, `DynamicUser=yes` and `RootDirectory=/some/dir`

> does it still apply to current git?

I think yes.

@tixxdz , what do you think?

**Comment by tixxdz** [comment]
@evverx @poettering yes it was about documentation (and may be incomplete code...), let me re-phrase:

We added lot of sandbox options that create implicitly a new mount namespace, however we never documented how these options behave if one of them is set and also `RootDirectory=` is set.

I relaxed the sandbox code a bit so our users can also use these options when `RootDirectory=` is set
https://github.com/systemd/systemd/blob/master/src/core/execute.c#L1982

If `RootDirectory` is set options like `ProtectHome=` `ProtectKernelModules=` `ProtectKernelTunables=`  will not fail hard if one of the protected path is not available inside the chroot.

The logic works for every path protected inside src/core/namespace.c that path can decide whether we should ignore it or no, that value takes precedence, however if that https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L98 is set to false then we check the global value:
https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L98

If `RootDirectory=...` set then we ignore these paths and we assume users have full control of their chroot and it's their fault if something fails, the only exception is `DynamicUser=` let it for later now.
We also have the `MountAPIVFS` that covers kernel virtual filesystems but only the root directories "/proc" "/dev" "/sys" , so maybe we should turn:

- [ ] to fix ?
https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L87
https://github.com/systemd/systemd/blob/master/src/core/namespace.c#L101
to `true` I'm not a fan of cgroupfs part of MountAPIVFS inside chroot hmm not sure about this...

- [ ] `RootDirectory=/...` works with `ProtectHome` and also with `ProtectSystem=yes` even if you do not have a /chroot/usr directory it's the user fault I guess. So this should be documented ?

- [ ] We have to test `RootDirectory=/..` and `ProtectSystem=full|strict`
    `ProtectSystem=full` should work even if you don't have /chroot/{usr|etc} inside but better test it and also document it ?

    `ProtectSystem=strict` should be fixed since it relates to /proc /sys and /dev we should add a check for it inside namespace_info_mount_apivfs() so that's an other thing to fix.


The only exception now when all this may fail is when `DynamicUser=` is set, if some paths are not there then we may fail, this was the reasoning at that time since DynamicUser= is a strong sandbox where we should be restricted to what we write to the filesystem due to UIG/GID recycling but also some of the options implied by `DynamicUser=` should not be ignored. Since there was the addition of `MountAPIVFS` may be we can go into the same direction and check if DynamicUser=yes should also work with `RootDirectory` ? otherwise we have to document it. 

These are the notes that I recall , I will check again this weekend, sorry was really busy :-/
